### DISCUSSION TITLE
Acccess CreateTreeNode in TreeNodesRenderingNotification

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#11425] [Features and ideas]
In Umbraco v8 with `TreeControllerBase.TreeNodesRendering` event we had access to `CreateTreeNode()` method, which was available via `sender`:

```
void TreeControllerBase_TreeNodesRendering(TreeControllerBase sender, TreeNodesRenderingEventArgs e)
{
    if (sender.TreeAlias == Vendr.Core.Constants.System.ProductAlias && e.QueryStrings["nodeType"] == Vendr.Core.Constants.Entities.EntityTypes.Store)
    {
        var mainRoute = Vendr.Umbraco.Constants.Sections.Commerce + "/" + Constants.System.ProductAlias;

        var storeId = e.QueryStrings["id"];
        var id = Constants.Trees.Reviews.Id;

        var reviewsNode = sender.CreateTreeNode(id, storeId, e.QueryStrings, "Reviews", Constants.Trees.Reviews.Icon, false, $"{mainRoute}/review-list/{storeId}");

        reviewsNode.Path = $"-1,{storeId},{id}";
        reviewsNode.NodeType = Constants.Trees.Reviews.NodeType;

        reviewsNode.AdditionalData.Add("storeId", storeId);
        reviewsNode.AdditionalData.Add("tree", Vendr.Umbraco.Constants.Trees.Stores.Alias);
        reviewsNode.AdditionalData.Add("application", Vendr.Umbraco.Constants.Sections.Commerce);

        var optNodeIndex = e.Nodes.FindIndex(x => x.NodeType == "Options");
        var index = optNodeIndex >= 0 ? optNodeIndex : e.Nodes.Count;

        e.Nodes.Insert(index, reviewsNode);
    }
}
```

This seems no longer to be the case with `TreeNodesRenderingNotification` notification in v9:

```
public class ReviewsTreeNodesNotification : INotificationHandler<TreeNodesRenderingNotification>
{
    private readonly UmbracoApiControllerTypeCollection _apiControllers;
    private readonly IUrlHelper _urlHelper;

    public ReviewsTreeNodesNotification(
        UmbracoApiControllerTypeCollection apiControllers,
        IUrlHelper urlHelper)
    {
        _apiControllers = apiControllers;
        _urlHelper = urlHelper;
    }

    public void Handle(TreeNodesRenderingNotification notification)
    {
        if (notification.TreeAlias == Vendr.Core.Constants.System.ProductAlias
            && notification.QueryString["nodeType"] == Vendr.Core.Constants.Entities.EntityTypes.Store)
        {
            var mainRoute = Vendr.Umbraco.Constants.Sections.Commerce + "/" + Constants.System.ProductAlias;

            var storeId = notification.QueryString["id"];
            var id = Constants.Trees.Reviews.Id;

            //var reviewsNode = CreateTreeNode(id, storeId, notification.QueryString, "Reviews", Constants.Trees.Reviews.Icon, false, $"{mainRoute}/review-list/{storeId}");

            string jsonUrl = _urlHelper.GetTreeUrl(_apiControllers, notification.GetType(), id, notification.QueryString);
            string menuUrl = _urlHelper.GetMenuUrl(_apiControllers, notification.GetType(), id, notification.QueryString);
            
            var reviewsNode = new TreeNode(id, storeId, jsonUrl, menuUrl)
            {
                Icon = Constants.Trees.Reviews.Icon,
                HasChildren = false,
                Name = "Reviews",
                Path = $"-1,{storeId},{id}",
                RoutePath = $"{mainRoute}/review-list/{storeId}",
                NodeType = Constants.Trees.Reviews.NodeType
            };

            reviewsNode.AdditionalData.Add("storeId", storeId);
            reviewsNode.AdditionalData.Add("tree", Vendr.Umbraco.Constants.Trees.Stores.Alias);
            reviewsNode.AdditionalData.Add("application", Vendr.Umbraco.Constants.Sections.Commerce);

            var optNodeIndex = notification.Nodes.FindIndex(x => x.NodeType == "Options");
            var index = optNodeIndex >= 0 ? optNodeIndex : notification.Nodes.Count;

            notification.Nodes.Insert(index, reviewsNode);
        }
    }
}
```

Instead we need to inject `UmbracoApiControllerTypeCollection` and `IUrlHelper` which has two extension methods `GetTreeUrl()` and `GetMenuUrl()`.
https://github.com/umbraco/Umbraco-CMS/blob/v9/contrib/src/Umbraco.Web.BackOffice/Trees/UrlHelperExtensions.cs#L44-L65

It would be great if we could access the `CreateTreeNode()` and maybe also `CreateRootNode()` like when creating a regular `TreeController`, which inherit these methods from `TreeControllerBase`:
https://github.com/umbraco/Umbraco-CMS/blob/v9/contrib/src/Umbraco.Web.BackOffice/Trees/TreeControllerBase.cs#L204-L218

In the example above it inject a tree node inject the existing Vendr tree.

Could we either inject a service or inherit these methods ... or access these from `TreeNodesRenderingNotification` parameter?

**Comment by bjarnef** [comment]
It seems when using `IUrlHelper` I get this error on startup:

![image](https://user-images.githubusercontent.com/2919859/138087636-e8b444c9-49d6-473b-95d2-bc63927bab5d.png)

If I register `IUrlHelper` like suggested in https://stackoverflow.com/a/54858702/1693918 the application is starting:

```
public static IUmbracoBuilder AddVendrReviews(this IUmbracoBuilder builder, Action<VendrReviewsSettings> defaultOptions = default)
{
    // Register configuration
    var options = builder.Services.AddOptions<VendrReviewsSettings>()
        .Bind(builder.Config.GetSection(Constants.System.ProductName));

    if (defaultOptions != default)
        options.Configure(defaultOptions);

    options.ValidateDataAnnotations();

    builder.Services.AddScoped<IUrlHelper>(factory =>
    {
        var actionContext = factory.GetService<IActionContextAccessor>()
                                   .ActionContext;
        return new UrlHelper(actionContext);
    });

    builder.Services.AddHttpContextAccessor();

    // Register services
    builder.Services.AddTransient<IReviewRepositoryFactory, ReviewRepositoryFactory>();
    builder.Services.AddSingleton<IReviewService, ReviewService>();
    builder.Services.AddSingleton<VendrReviewsApi>();

    // Register event handlers
    builder.AddNotificationHandler<TreeNodesRenderingNotification, ReviewsTreeNodesNotification>();

    // Register component
    builder.Components()
        .Append<VendrReviewsComponent>();

    return builder;
}
```

But then cause error when logging into backoffice:
![image](https://user-images.githubusercontent.com/2919859/138088483-38c5f087-2540-4a5e-bb43-247eec80a885.png)

@bergmania is there something we should register to use `IUrlHelper` to use the `GetTreeUrl()` and `GetMenu()` extensions?

**Comment by bjarnef** [comment]
It found a similar example here:
https://github.com/Dotnet-Boxed/Framework/issues/18#issuecomment-379360168

```
services.AddScoped<IUrlHelper>(factory =>
{
       var actionContext = factory.GetService<IActionContextAccessor>().ActionContext;
       return new UrlHelper(actionContext);
});
```

**Comment by bjarnef** [comment]
This inject `IActionContextAccessor` which seems to create the tree node. although not correct and with some errors.
I guess it is missing something in the action context.

```
public class ReviewsTreeNodesNotification : INotificationHandler<TreeNodesRenderingNotification>
{
    private readonly UmbracoApiControllerTypeCollection _apiControllers;
    private readonly IActionContextAccessor _actionContextAccessor;
    //private readonly IUrlHelper _urlHelper;

    public ReviewsTreeNodesNotification(
        UmbracoApiControllerTypeCollection apiControllers,
        IActionContextAccessor actionContextAccessor
        //IUrlHelper urlHelper
        )
    {
        _apiControllers = apiControllers;
        _actionContextAccessor = actionContextAccessor;
        //_urlHelper = urlHelper;
    }

    public void Handle(TreeNodesRenderingNotification notification)
    {
        if (notification.TreeAlias == Vendr.Core.Constants.System.ProductAlias
            && notification.QueryString["nodeType"] == Vendr.Core.Constants.Entities.EntityTypes.Store)
        {
            var mainRoute = Vendr.Umbraco.Constants.Sections.Commerce + "/" + Constants.System.ProductAlias;

            var storeId = notification.QueryString["id"];
            var id = Constants.Trees.Reviews.Id;

            //var reviewsNode = CreateTreeNode(id, storeId, notification.QueryString, "Reviews", Constants.Trees.Reviews.Icon, false, $"{mainRoute}/review-list/{storeId}");

            var urlHelper = new Microsoft.AspNetCore.Mvc.Routing.UrlHelper(_actionContextAccessor.ActionContext);

            string jsonUrl = urlHelper.GetTreeUrl(_apiControllers, notification.GetType(), id, notification.QueryString);
            string menuUrl = urlHelper.GetMenuUrl(_apiControllers, notification.GetType(), id, notification.QueryString);
            
            var reviewsNode = new TreeNode(id, storeId, jsonUrl, menuUrl)
            {
                Icon = Constants.Trees.Reviews.Icon,
                HasChildren = false,
                Name = "Reviews",
                Path = $"-1,{storeId},{id}",
                RoutePath = $"{mainRoute}/review-list/{storeId}",
                NodeType = Constants.Trees.Reviews.NodeType
            };

            reviewsNode.AdditionalData.Add("storeId", storeId);
            reviewsNode.AdditionalData.Add("tree", Vendr.Umbraco.Constants.Trees.Stores.Alias);
            reviewsNode.AdditionalData.Add("application", Vendr.Umbraco.Constants.Sections.Commerce);

            var optNodeIndex = notification.Nodes.FindIndex(x => x.NodeType == "Options");
            var index = optNodeIndex >= 0 ? optNodeIndex : notification.Nodes.Count;

            notification.Nodes.Insert(index, reviewsNode);
        }
    }
}
```

Maybe there's a workaround to get an existing tree (in this case Vendr tree).

**Reply by bergmania** [reply]
Sounds okay to me, if the `TreeNodesRenderingNotification` had an func you could execute:

 ```cs
       public TreeNodesRenderingNotification(TreeNodeCollection nodes, FormCollection queryString, string treeAlias, Func<string, string, FormCollection, string, TreeNode> createTreeNode)
        {
            Nodes = nodes;
            QueryString = queryString;
            TreeAlias = treeAlias;
            CreateTreeNode = createTreeNode;
        }
```

And then publish would need to be something like this:
```cs
await _eventAggregator.PublishAsync(new TreeNodesRenderingNotification(nodes, queryStrings, TreeAlias, CreateTreeNode));
```

**Reply by bjarnef** [reply]
Yes, that would be great.
Alternatively an service to inject, which contained a `CreateTreeNode()` function like in `TreeControllerBase` class.

I tried the following workaround, but I think it doesn't quite know the context as the "Reviews" node is create at root and not under Vendr store node.
https://github.com/vendrcontrib/vendr-reviews/blob/v2/dev/src/Vendr.Contrib.Reviews/Notifications/ReviewsTreeNodesNotification.cs#L43-L56

![image](https://user-images.githubusercontent.com/2919859/138121647-11195d28-7904-4721-9959-6dbf72786111.png)

Feel free to try it out on Umbraco V9, Vendr 2.0+ and the `v2/dev` in Vendr Reviews repository.
or just use the `ReviewsTreeNodesNotification.cs` example here (should be possible to reproduce by injecting the tree node in any other existing trees):
https://github.com/vendrcontrib/vendr-reviews/blob/v2/dev/src/Vendr.Contrib.Reviews/Notifications/ReviewsTreeNodesNotification.cs

**Reply by bjarnef** [reply]
I tried injecting `IUrlHelperFactory` in https://github.com/vendrcontrib/vendr-reviews/commit/5187e0116627bfbce823f10ebe33259917e598ec

But it also required adding `services.AddSingleton<IActionContextAccessor, ActionContextAccessor>();` in Startup.cs

However I am not sure why it render the other "Reviews" root node.

![image](https://user-images.githubusercontent.com/2919859/138134143-59d966fc-ee68-4534-8f3a-cfb5c7270b54.png)

**Reply by bjarnef** [reply]
I could inject `IUrlHelperFactory` and `IActionContextAccessor`.
Then I have this:

```
var urlHelper = _urlHelperFactory.GetUrlHelper(_actionContextAccessor.ActionContext);

string jsonUrl = urlHelper.GetTreeUrl(_apiControllers, typeof(StoresTreeController), id, notification.QueryString);
string menuUrl = urlHelper.GetMenuUrl(_apiControllers, typeof(StoresTreeController), id, notification.QueryString);
                
var reviewsNode = new TreeNode(id, storeId, jsonUrl, menuUrl)
{
        Icon = Constants.Trees.Reviews.Icon,
        HasChildren = false,
        Name = "Reviews",
        Path = $"-1,{storeId},{id}",
        RoutePath = $"{mainRoute}/review-list/{storeId}",
        NodeType = Constants.Trees.Reviews.NodeType
};

reviewsNode.AdditionalData.Add("storeId", storeId);
reviewsNode.AdditionalData.Add("tree", Vendr.Umbraco.Constants.Trees.Stores.Alias);
reviewsNode.AdditionalData.Add("application", Vendr.Umbraco.Constants.Sections.Commerce);

var optNodeIndex = notification.Nodes.FindIndex(x => x.NodeType == "Options");
var index = optNodeIndex >= 0 ? optNodeIndex : notification.Nodes.Count;

notification.Nodes.Insert(index, reviewsNode);
```

where `StoresTreeController` is `Vendr.Umbraco.Web.Trees.StoresTreeController`.

However it seems to mix the trees from settings and commerce section somehow as in screenshot above.
There should be a "Stores" node at root in Commerce section - only each store node at root.

Furthermore I am not sure why the "Third Party" tree group is rendered.

**Reply by bjarnef** [reply]
@bergmania do you know if there a workaround to solve this for now or do we need some core changes? Currently this is a showstooper to make Vendr Reviews work on Umbraco 9 as I need to inject the tree node in the Vendr tree:
https://github.com/vendrcontrib/vendr-reviews/blob/v2/dev/src/Vendr.Contrib.Reviews/Notifications/ReviewsTreeNodesNotification.cs

**Reply by bergmania** [reply]
I think you can just do like you say directly in the notification handler, and implement the CreateTreeNode directly.. Something like this:

```cs
 public class TreeNodesRenderingNotificationHandler : INotificationHandler<TreeNodesRenderingNotification>
    {
        private readonly IUrlHelperFactory _urlHelperFactory;
        private readonly IActionContextAccessor _actionContextAccessor;
        private readonly UmbracoApiControllerTypeCollection _apiControllers;

        public TreeNodesRenderingNotificationHandler(IUrlHelperFactory urlHelperFactory, IActionContextAccessor actionContextAccessor, UmbracoApiControllerTypeCollection apiControllers)
        {
            _urlHelperFactory = urlHelperFactory;
            _actionContextAccessor = actionContextAccessor;
            _apiControllers = apiControllers;
        }

        public TreeNode CreateTreeNode(string id, string parentId, FormCollection queryStrings, string title, string icon, string routePath)
        {
            var urlHelper = _urlHelperFactory.GetUrlHelper(_actionContextAccessor.ActionContext);

            var controllerActionDescriptor =
                _actionContextAccessor.ActionContext.ActionDescriptor as ControllerActionDescriptor;

            var jsonUrl = urlHelper.GetTreeUrl(_apiControllers, controllerActionDescriptor.ControllerTypeInfo, id, queryStrings);
            var menuUrl = urlHelper.GetMenuUrl(_apiControllers, controllerActionDescriptor.ControllerTypeInfo, id, queryStrings);
            var node = new TreeNode(id, parentId, jsonUrl, menuUrl) { Name = title, RoutePath = routePath, Icon = icon };
            return node;
        }

        public void Handle(TreeNodesRenderingNotification notification)
        {
             //TODO implement where you use CreateTreeNode
        }
    }
```

**Reply by bjarnef** [reply]
Thanks for the example @bergmania
It seems to work in same way as when using `typeof(StoresController)`:

```
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Controllers;
using Microsoft.AspNetCore.Mvc.Infrastructure;
using Microsoft.AspNetCore.Mvc.Routing;
using Umbraco.Cms.Core;
using Umbraco.Cms.Core.Events;
using Umbraco.Cms.Core.Security;
using Umbraco.Cms.Core.Notifications;
using Umbraco.Cms.Core.Trees;
using Umbraco.Cms.Core.Web;
using Umbraco.Cms.Web.BackOffice.Trees;
using Vendr.Umbraco.Web.Trees;

namespace Vendr.Contrib.Reviews.Notifications
{
    public class ReviewsTreeNodesNotification : INotificationHandler<TreeNodesRenderingNotification>
    {
        private readonly UmbracoApiControllerTypeCollection _apiControllers;
        private readonly IUrlHelperFactory _urlHelperFactory;
        private readonly IActionContextAccessor _actionContextAccessor;
        //private readonly IUrlHelper _urlHelper;

        public ReviewsTreeNodesNotification(
            UmbracoApiControllerTypeCollection apiControllers,
            IUrlHelperFactory urlHelperFactory,
            IActionContextAccessor actionContextAccessor
            //IUrlHelper urlHelper
            )
        {
            _apiControllers = apiControllers;
            _urlHelperFactory = urlHelperFactory;
            _actionContextAccessor = actionContextAccessor;
            //_urlHelper = urlHelper;
        }

        public TreeNode CreateTreeNode(string id, string parentId, FormCollection queryStrings, string title, string icon, string routePath)
        {
            var urlHelper = _urlHelperFactory.GetUrlHelper(_actionContextAccessor.ActionContext);

            var controllerActionDescriptor =
                _actionContextAccessor.ActionContext.ActionDescriptor as ControllerActionDescriptor;

            var jsonUrl = urlHelper.GetTreeUrl(_apiControllers, controllerActionDescriptor.ControllerTypeInfo, id, queryStrings);
            var menuUrl = urlHelper.GetMenuUrl(_apiControllers, controllerActionDescriptor.ControllerTypeInfo, id, queryStrings);

            var node = new TreeNode(id, parentId, jsonUrl, menuUrl) { Name = title, RoutePath = routePath, Icon = icon };
            return node;
        }

        public void Handle(TreeNodesRenderingNotification notification)
        {
            if (notification.TreeAlias == Vendr.Core.Constants.System.ProductAlias
                && notification.QueryString["nodeType"] == Vendr.Core.Constants.Entities.EntityTypes.Store)
            {
                var mainRoute = Vendr.Umbraco.Constants.Sections.Commerce + "/" + Constants.System.ProductAlias;

                var storeId = notification.QueryString["id"];
                var id = Constants.Trees.Reviews.Id;

                //var reviewsNode = CreateTreeNode(id, storeId, notification.QueryString, "Reviews", Constants.Trees.Reviews.Icon, false, $"{mainRoute}/review-list/{storeId}");

                //var urlHelper = _urlHelperFactory.GetUrlHelper(_actionContextAccessor.ActionContext);
                //var urlHelper = new Microsoft.AspNetCore.Mvc.Routing.UrlHelper(_actionContextAccessor.ActionContext);

                //string jsonUrl = urlHelper.GetTreeUrl(_apiControllers, typeof(StoresTreeController), id, notification.QueryString);
                //string menuUrl = urlHelper.GetMenuUrl(_apiControllers, typeof(StoresTreeController), id, notification.QueryString);
                
                //var reviewsNode = new TreeNode(id, storeId, jsonUrl, menuUrl)
                //{
                //    Icon = Constants.Trees.Reviews.Icon,
                //    HasChildren = false,
                //    Name = "Reviews",
                //    Path = $"-1,{storeId},{id}",
                //    RoutePath = $"{mainRoute}/review-list/{storeId}",
                //    NodeType = Constants.Trees.Reviews.NodeType
                //};

                var reviewsNode = CreateTreeNode(id, storeId, notification.QueryString, "Reviews", Constants.Trees.Reviews.Icon, $"{mainRoute}/review-list/{storeId}");

                reviewsNode.Path = $"-1,{storeId},{id}";
                reviewsNode.NodeType = Constants.Trees.Reviews.NodeType;

                reviewsNode.AdditionalData.Add("storeId", storeId);
                reviewsNode.AdditionalData.Add("tree", Vendr.Umbraco.Constants.Trees.Stores.Alias);
                reviewsNode.AdditionalData.Add("application", Vendr.Umbraco.Constants.Sections.Commerce);

                var optNodeIndex = notification.Nodes.FindIndex(x => x.NodeType == "Options");
                var index = optNodeIndex >= 0 ? optNodeIndex : notification.Nodes.Count;

                notification.Nodes.Insert(index, reviewsNode);
            }
        }
    }
}
```

The "Reviews" node is injected, but when clicking the node it request `/umbraco/views/vendrReviews/review-list.html`.
I would expect it to request `/App_Plugins/VendrReviews/backoffice/views/review-list.html` similar to e.g. Vendr request views like this: `/App_Plugins/Vendr/backoffice/views/discount/list.html`

Not sure if something specific regarding routing has changed from v8 to v9 here?

Furthermore I also wonder what is rendering the "Third Party" tree group 🤔
![image](https://user-images.githubusercontent.com/2919859/138700037-e3f52ed2-b9c0-48a4-9290-75da00c73b35.png)

**Reply by bjarnef** [reply]
Maybe it doesn't respect `Tree` attribute here:
https://github.com/vendrcontrib/vendr-reviews/blob/v2/dev/src/Vendr.Contrib.Reviews/Web/Controllers/ReviewTreeController.cs#L48-L53

I tried specific adding `TreeGroup = "commerce"`, but seems to render same result.

**Reply by bergmania** [reply]
The "Third Party" group should only be used if TreeGroup is null or empty. 
If I use `TreeGroup = "commerce"` there are created two trees which I guess is expected. if I use `TreeGroup = "commerceSettingsGroup"` then it is added to the existing Commerce-group. 

I think the reason you see the `/umbraco/views/vendrReviews/review-list.html` url is because it cannot find a tree alias with `vendrReviews`.

I think this `var mainRoute = Vendr.Umbraco.Constants.Sections.Commerce + "/" + Constants.System.ProductAlias;` should be `var mainRoute = Vendr.Umbraco.Constants.Sections.Commerce + "/" + Constants.Trees.Reviews.Alias;`.

**Reply by bjarnef** [reply]
In Vendr Reviews for v8 I didn't set `TreeGroup` and it didn't added the "Third Party" group. I tried use `TreeGroup` in the tree controller, but it didn't seem to make a difference.

I would expect `TreeGroup = "commerceSettingsGroup"` if I wanted to inject the tree in the Vendr Settings.. but in this case it should inject the tree in the Vendr tree in Commerce section.

Regarding `mainRoute` it was `commerce/vendrreviews` in v1:
https://github.com/vendrcontrib/vendr-reviews/blob/v1/main/src/Vendr.Contrib.Reviews/Components/VendrReviewsComponent.cs#L33

The `ProductAlias` constant should be `vendrReviews` as well in v2:
https://github.com/vendrcontrib/vendr-reviews/blob/b445e6df336049981cdc3dacb5fef7a6dbfaf9c0/src/Vendr.Contrib.Reviews/Constants.cs#L15

So it should be `commerce/vendrreviews` as well, but not sure if the something in the routing has changed in Umbraco 9 regarding this?

**Reply by bjarnef** [reply]
I tried `var mainRoute = Vendr.Umbraco.Constants.Sections.Commerce + "/" + Constants.Trees.Reviews.Alias;` which request the following instead:
/App_Plugins/VendrReviews/backoffice/reviews/review-list.html

Although it still cause an error... not sure why the route need to be different in v9 though.

**Reply by bjarnef** [reply]
I tried with `TreeGroup = "commerce"` but didn't make a different from not set `TreeGroup`.

If I set `TreeGroup = "commerceSettingsGroup"` the "Third Party" tree group is not shown, but it still add a weird tree node here:

![image](https://user-images.githubusercontent.com/2919859/138838996-8109c604-3f79-4cef-959b-89f9f4141b0d.png)

If I try to expand this node I get these errors:

![image](https://user-images.githubusercontent.com/2919859/138839360-09c00d79-d03a-4bc4-8346-78031ccc6d75.png)

**Reply by bjarnef** [reply]
Okay, the `mainRoute` was correct as it was, but apparently I missed to copy the `package.manifest` to the Vendr demo store, where I tested this. Now it looks like this:

![image](https://user-images.githubusercontent.com/2919859/138842956-1ff3b6fa-d571-4b1d-b90e-4ccae383ba29.png)

**Reply by bjarnef** [reply]
Maybe it is missing something, which is handled in `TreeControllerBase` or something might has changed regarding `TreeGroup` from v8 to v9?

**Reply by bergmania** [reply]
Not sure, there shouldn't be changes as far as Im aware.. And couldn't find differences when I just briefly compared the tree implementations in v8 and v9

**Reply by bjarnef** [reply]
Just a guess, could the UrlHelper or ActionContext be missing any route data here, which `TreeControllerBase` handle?

**Reply by bjarnef** [reply]
@bergmania okay, so this in weird. In seems `TreeUse.None` instead of `TreeUse.Main` works.

![image](https://user-images.githubusercontent.com/2919859/138881739-085359a9-0d28-4f48-9531-f3c60bc86beb.png)

However v1 of Vendr Review did use `TreeUse.Main`, so strange this is different in v9.
From the comment I would have though the `TreeUse.None` should be used if you didn't want to render the tree, hide the tree or maybe not had a view for the tree like the "Options" tree menu in Vendr.

![image](https://user-images.githubusercontent.com/2919859/138882104-236bd894-f74a-4de8-955d-145a2982392f.png)
